### PR TITLE
Store errors on single changes

### PR DESCRIPTION
**Pull request by rebstar6**
Just storage in this PR. Still todo in future PRs:
- Deal with response serialization (JsonProtocol)
- Bring in an actual soft failure

### CODE REVIEW

File: modules/core/src/main/scala/vinyldns/core/domain/DomainValidationErrors.scala [hunk 1]
```diff
@@ -24,6 +24,17 @@ sealed abstract class DomainValidationError(val isFatal: Boolean = true) {
   def message: String
 }
 
+object DomainValidationErrorType extends Enumeration {
+  type DomainValidationErrorType = Value
+  // NOTE: once defined, an error code cannot be changed!
+  val ZDE, Unknown = Value
+
+  def from(error: DomainValidationError): DomainValidationErrorType = error match {
```
Comment (mitruly) [inline modules/core/src/main/scala/vinyldns/core/domain/DomainValidationErrors.scala]: I think I'm fine with having the type used. May be worthwhile to store the error message string at the time that the error was associated with the `SingleChange`.

I don't think it's worth trying to store parameters for the error types since some of our error input parameters have changed in the past, and this would require migration scripts each time these are updated.

File: modules/core/src/main/scala/vinyldns/core/domain/DomainValidationStoredError.scala [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+object DomainValidationStoredError {
+  def apply(error: DomainValidationError): DomainValidationStoredError =
+    new DomainValidationStoredError(DomainValidationErrorType.from(error), error.message)
+}
+
+object DomainValidationErrorType extends Enumeration {
+  type DomainValidationErrorType = Value
+  // NOTE: once defined, an error code cannot be changed!
+  val ZDE, Unknown = Value
```
Comment (rebstar6) [inline modules/core/src/main/scala/vinyldns/core/domain/DomainValidationStoredError.scala]: will build this out with all the types rather than an unknown

File: modules/core/src/main/scala/vinyldns/core/protobuf/BatchChangeProtobufConversions.scala [hunk 3]
```diff
@@ -117,21 +122,37 @@ trait BatchChangeProtobufConversions extends ProtobufConversions {
...
-      change.recordName.foreach(x => sc.setRecordName(x))
-      change.systemMessage.foreach(x => sc.setSystemMessage(x))
-      change.recordChangeId.foreach(x => sc.setRecordChangeId(x))
-      change.recordSetId.foreach(x => sc.setRecordSetId(x))
 
+      addCommonSingleChangeFields(sc, change)
       sc.build()
     }
 
+  def addCommonSingleChangeFields(
```
Comment (rebstar6) [inline modules/core/src/main/scala/vinyldns/core/protobuf/BatchChangeProtobufConversions.scala]: consolidating. unit because these functions are actually modifying the builder itself

File: modules/core/src/main/scala/vinyldns/core/domain/DomainValidationStoredError.scala [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  // NOTE: once defined, an error code type cannot be changed!
+  val ChangeLimitExceeded, BatchChangeIsEmpty, GroupDoesNotExist, NotAMemberOfOwnerGroup,
+  InvalidDomainName, InvalidLength, InvalidEmail, InvalidRecordType, InvalidPortNumber,
+  InvalidIpv4Address, InvalidIpv6Address, InvalidIPAddress, InvalidTTL, InvalidMxPreference,
+  InvalidBatchRecordType, ZoneDiscoveryError, RecordAlreadyExists, RecordDoesNotExist,
+  CnameIsNotUniqueError, UserIsNotAuthorized, RecordNameNotUniqueInBatch, RecordInReverseZoneError,
+  HighValueDomainError, MissingOwnerGroupId, ExistingMultiRecordError, NewMultiRecordError,
+  CnameAtZoneApexError = Value
+
+  def from(error: DomainValidationError): DomainValidationErrorType =
```
Comment (rebstar6) [inline modules/core/src/main/scala/vinyldns/core/domain/DomainValidationStoredError.scala]: at a certain point, does it make sense to just store `error.getClass.getSimpleName` as a string and not enumerate? its less safe in the sense that DomainValidationStoredError would then just be strings, but also this feels perhaps silly

File: modules/core/src/main/scala/vinyldns/core/domain/DomainValidationStoredError.scala [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package vinyldns.core.domain
+
+import vinyldns.core.domain.DomainValidationErrorType.DomainValidationErrorType
+
+case class DomainValidationStoredError(errorType: DomainValidationErrorType, message: String)
```
Comment (pauljamescleary) [inline modules/core/src/main/scala/vinyldns/core/domain/DomainValidationStoredError.scala]: I know naming is hard, but "Stored" feels inconsequential to the domain, our implementation leaking into the domain model.  Imagine communicating this to Dave Back or Joe Crowe, what is a "DomainValidationStoredError", what words would be more suitable?

We are indicating "these are errors that happened as part of evaluation that need to be fixed".  There is also this "temporary" nature to them as they can be resolved and essentially disappear.

We are also conflating the set of "all validation errors" with "those that may need resolution", the latter is a small subset of the former.

Coming up with an alternative name is not easy as things conflate, "BatchChangeIssue" or "BatchChangeProblem" or "BatchChangeHalt" come to mind.

File: modules/core/src/main/scala/vinyldns/core/protobuf/BatchChangeProtobufConversions.scala [hunk 6]
```diff
@@ -73,18 +78,28 @@ trait BatchChangeProtobufConversions extends ProtobufConversions {
...
             change.getId
           )
-
       }
     }
 
+  def fromPB(errors: List[VinylDNSProto.DomainValidationStoredError])
+    : Either[Throwable, List[DomainValidationStoredError]] = Either.catchNonFatal {
+    errors.map { e =>
+      val errorType = DomainValidationErrorType.withName(e.getErrorType)
```
Comment (pauljamescleary) [inline modules/core/src/main/scala/vinyldns/core/protobuf/BatchChangeProtobufConversions.scala]: I think `withName` will throw if there is not a match

File: modules/core/src/main/scala/vinyldns/core/domain/DomainValidationStoredError.scala (hunk 2 above)
Comment (pauljamescleary) [reply modules/core/src/main/scala/vinyldns/core/domain/DomainValidationStoredError.scala]: not in love with `ZDE`, as TLAs naturally diverge from some standard and you get FLAs or variations thereof

File: modules/core/src/main/scala/vinyldns/core/domain/DomainValidationStoredError.scala (hunk 5 above)
Comment (pauljamescleary) [reply modules/core/src/main/scala/vinyldns/core/domain/DomainValidationStoredError.scala]: Also "FailureReason"

Comment (rebstar6) [reply modules/core/src/main/scala/vinyldns/core/domain/DomainValidationStoredError.scala]: none of this is yet hitting the user as we specify the SingleAddChangeSerializer/SingleDeleteChangeSerializer rather than relying on any default values, so just keep that in mind. In either case, I will change.

File: modules/core/src/main/scala/vinyldns/core/protobuf/BatchChangeProtobufConversions.scala (hunk 6 above)
Comment (rebstar6) [reply modules/core/src/main/scala/vinyldns/core/protobuf/BatchChangeProtobufConversions.scala]: wont that get caught by the `Either.catchNonFatal`?

File: modules/api/src/main/scala/vinyldns/api/domain/record/RecordSetValidations.scala [hunk 7]
```diff
@@ -126,7 +126,7 @@ object RecordSetValidations {
       ensuring(
         RecordSetAlreadyExists(s"RecordSet with name ${newRecordSet.name} already " +
           s"exists in zone ${zone.name}, CNAME record cannot use duplicate name"))(
-        !existingRecordsWithName.exists(_.id != newRecordSet.id)
+        existingRecordsWithName.forall(_.id == newRecordSet.id)
```
Comment (rebstar6) [inline modules/api/src/main/scala/vinyldns/api/domain/record/RecordSetValidations.scala]: this is unrelated, i was looking at it and the double negative confused me and i forgot i changed it until i saw it here

File: modules/core/src/main/scala/vinyldns/core/domain/DomainValidationErrors.scala [hunk 8]
```diff
@@ -97,7 +97,8 @@ final case class InvalidBatchRecordType(param: String, supported: Set[RecordType
     s"""Invalid Batch Record Type: "$param", valid record types for batch changes include $supported."""
 }
 
-final case class ZoneDiscoveryError(name: String) extends DomainValidationError {
+final case class ZoneDiscoveryError(name: String, fatal: Boolean = true)
```
Comment (rebstar6) [inline modules/core/src/main/scala/vinyldns/core/domain/DomainValidationErrors.scala]: still leaving at fatal, needed to specify for testing (cant mock a final case class, and DomainValidationError is a sealed trait...means i cant mock out something that will pass the match in DomainValidationErrorType

Comment (rebstar6) [reply modules/core/src/main/scala/vinyldns/core/domain/DomainValidationErrors.scala]: mocking the fatal response back. was just explaining why this change was needed

**Comment by britneywright** [comment]
Looks like there's only test coverage on ZoneDiscoveryError and HighValueDomain error in the SingleChangeError `from` method, so coverage of that file is only 13.97%
